async-queue/src/fifo.rs

```diff
@@ -0,0 +1,403 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use super::*;
+use fortanix_sgx_abi::{FifoDescriptor, Return, Usercall};
+use lazy_static::lazy_static;
+use std::collections::HashSet;
+use std::mem;
+use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
+use std::sync::Mutex;
+
+impl<T: WithAtomicId> Fifo<T> {
+    /// panics if len is not a power of two.
+    pub fn new(len: usize) -> Self {
+        assert!(len.is_power_of_two(), "Fifo len should be a power of two");
+        let mut data = Vec::with_capacity(len);
+        data.resize_with(len, T::empty);
+        Self {
+            data: data.into_boxed_slice(),
+            offsets: Box::new(AtomicUsize::new(0)),
+        }
+    }
+
+    pub fn descriptor(&mut self) -> FifoDescriptor<T> {
+        FifoDescriptor {
+            data: self.data.as_mut().as_mut_ptr(),
+            len: self.data.len(),
+            offsets: self.offsets.as_ref() as _,
+        }
+    }
+
+    pub fn sender<S: Synchronizer>(&mut self, synchronizer: S) -> Sender<T, S> {
+        Sender::new(self.descriptor(), synchronizer)
+    }
+
+    /// Panics if there is an existing (sync or async) receiver for the same queue.
+    pub fn receiver<S: Synchronizer>(&mut self, synchronizer: S) -> Receiver<T, S> {
+        Receiver::new(self.descriptor(), synchronizer)
+    }
+
+    pub fn async_sender<S: AsyncSynchronizer>(&mut self, synchronizer: S) -> AsyncSender<T, S> {
+        AsyncSender::new(self.descriptor(), synchronizer)
+    }
+
+    /// Panics if there is an existing (sync or async) receiver for the same queue.
+    pub fn async_receiver<S: AsyncSynchronizer>(&mut self, synchronizer: S) -> AsyncReceiver<T, S> {
+        AsyncReceiver::new(self.descriptor(), synchronizer)
+    }
```

Comment:
These APIs seem not that useful to me. If you want to try to create a safer API, consider adding lifetimes to Sender/Receiver (e.g. using PhantomData). Also, you should be able to create a channel `(Sender, Receiver)` in a single call taking `&mut self`. You should then be able to convert one of them (taken by value) into a FifoDescriptor.